From af8f3d23a90bd6904baf1bc071df039af1aa551d Mon Sep 17 00:00:00 2001
From: "T.J. Mercier" <tjmercier@google.com>
Date: Mon, 20 Apr 2026 09:57:53 -0700
Subject: [PATCH 113/223] ANDROID: Introduce get_dmabuf_info()

Instead of open coding the refcount management for task_dma_buf_infos,
provide a new function get_dmabuf_info() that acquires a reference in a
similar way that put_dmabuf_info() releases one.

Bug: 496996374
Change-Id: I41c4d204ad9500f818be2cb2285c06a3beae6a71
Signed-off-by: T.J. Mercier <tjmercier@google.com>
(cherry picked from commit 00cb18c67d49f5d991e032e538572762834ccaec)
Signed-off-by: Lee Jones <joneslee@google.com>
---
 drivers/dma-buf/dma-buf.c | 13 +++++++++----
 fs/exec.c                 |  3 ++-
 include/linux/dma-buf.h   |  2 ++
 3 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/drivers/dma-buf/dma-buf.c b/drivers/dma-buf/dma-buf.c
index 818b78374e90..d08b98a3ffc0 100644
--- a/drivers/dma-buf/dma-buf.c
+++ b/drivers/dma-buf/dma-buf.c
@@ -640,7 +640,7 @@ int copy_dmabuf_info(u64 clone_flags, struct task_struct *task)
 	 * the parent, so they can both share the same dmabuf_info.
 	 */
 	if (share_vm && share_fs) {
-		refcount_inc(&parent_dmabuf_info->refcnt);
+		get_dmabuf_info(parent_dmabuf_info);
 		task->dmabuf_info = parent_dmabuf_info;
 
 		if (task->mm) {
@@ -669,17 +669,22 @@ int copy_dmabuf_info(u64 clone_flags, struct task_struct *task)
 	task->dmabuf_info = child_dmabuf_info;
 
 	if (task->mm) {
-		refcount_inc(&child_dmabuf_info->refcnt);
+		get_dmabuf_info(child_dmabuf_info);
 		task->mm->dmabuf_info = child_dmabuf_info;
 	}
 	if (task->files) {
-		refcount_inc(&child_dmabuf_info->refcnt);
+		get_dmabuf_info(child_dmabuf_info);
 		task->files->dmabuf_info = child_dmabuf_info;
 	}
 
 	return 0;
 }
 
+void get_dmabuf_info(struct task_dma_buf_info *dmabuf_info)
+{
+	refcount_inc(&dmabuf_info->refcnt);
+}
+
 void put_dmabuf_info(struct task_dma_buf_info *dmabuf_info)
 {
 	if (!dmabuf_info)
@@ -782,7 +787,7 @@ int dma_buf_begin_new_exec(struct files_struct *old_files)
 			put_dmabuf_info(my_files->dmabuf_info);
 
 		/* Finally swap over to the new dmabuf info */
-		refcount_inc(&new_dmabuf_info->refcnt);
+		get_dmabuf_info(new_dmabuf_info);
 		my_files->dmabuf_info = new_dmabuf_info;
 		spin_unlock(&my_files->file_lock);
 
diff --git a/fs/exec.c b/fs/exec.c
index c322ff1fffd8..965fdc593d65 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1307,7 +1307,8 @@ int begin_new_exec(struct linux_binprm * bprm)
 	 * dma_buf_begin_new_exec) for the new mm_struct.
 	 */
 	if (IS_ENABLED(CONFIG_DMA_SHARED_BUFFER)) {
-		refcount_inc(&current->dmabuf_info->refcnt);
+		if (current->dmabuf_info)
+			get_dmabuf_info(current->dmabuf_info);
 		me->mm->dmabuf_info = current->dmabuf_info;
 	}
 
diff --git a/include/linux/dma-buf.h b/include/linux/dma-buf.h
index c90e9a181301..196fcac1e205 100644
--- a/include/linux/dma-buf.h
+++ b/include/linux/dma-buf.h
@@ -806,6 +806,7 @@ int is_dma_buf_file(struct file *file);
 int dma_buf_account_task(struct dma_buf *dmabuf, struct task_dma_buf_info *dmabuf_info);
 void dma_buf_unaccount_task(struct dma_buf *dmabuf, struct task_dma_buf_info *dmabuf_info);
 int copy_dmabuf_info(u64 clone_flags, struct task_struct *task);
+void get_dmabuf_info(struct task_dma_buf_info *dmabuf_info);
 void put_dmabuf_info(struct task_dma_buf_info *dmabuf_info);
 int dma_buf_begin_new_exec(struct files_struct *old_files);
 
@@ -818,6 +819,7 @@ static inline void dma_buf_unaccount_task(struct dma_buf *dmabuf,
 					  struct task_dma_buf_info *dmabuf_info) {}
 static inline int copy_dmabuf_info(u64 clone_flags,
 				   struct task_struct *task) { return 0; }
+static inline void get_dmabuf_info(struct task_dma_buf_info *dmabuf_info) {}
 static inline void put_dmabuf_info(struct task_dma_buf_info *dmabuf_info) {}
 static inline int dma_buf_begin_new_exec(struct files_struct *old_files) { return 0; }
 
-- 
2.50.1

